text/0000-introduce-a-new-signing-mechanism.md

```diff
@@ -0,0 +1,558 @@
+
+1. Submit the signature and public signing certificate to Rekor before the signing certificate expires (example uses pseudocode):
+
+   ```
+   POST https://rekor.sigstore.dev/api/v1/log/entries
+   {
+    kind: "rekord",
+    apiVersion: "0.0.1",
+    spec: {
+      signature: {
+        format: "x509",
+        content: Base64.encode64(signature),
+        publicKey: {
+          content: Base64.encode64(cert_chain_pem),
+        },
+      },
+      data: {
+        content: Base64.encode64(gem_contents),
+        hash: {
+          algorithm: "sha256",
+          value: gem_digest,
+        },
+      },
+    },
+   }
+   ```
+
+The gem signature and certificate chain are stored together in an immutable Rekor log entry, such as https://rekor.sigstore.dev/api/v1/log/entries/e66b75a340d2cf240f92cc0fd7f77c3ea0855c7eb5e3e9e2ae873567d9491fdc. In the signature verification flow, we will be able to locate this entry by searching Rekor using the .gem file's digest.
+
+### The verifying flow
+
+The signature verification flow occurs at gem installation time. In our proposed Phase 1, verification is opt-in, and the signature entry is only hosted on Rekor itself. By Phase 2, we anticipate that rubygems.org will keep a copy of the signature as well, and that clients will download it alongside the gem itself, decoupling the install experience from any Rekor queries on the happy path.
+
+The following examples illustrate Phase 1.
+
+<img src="https://user-images.githubusercontent.com/33674553/151441714-5b69f814-39b4-41e1-9dbc-b8241dfb83b6.png" width="45%"/>
+
+(Click on image for full-sized version).
+
+1. Query Rekor for a list of log entry UUIDs for a single gem file digest. 
+   
+   ```
+   POST https://rekor.sigstore.dev/api/v1/index/retrieve
+   {
+     hash: "sha256:b65e5c01770ae7d46293540cc20dd3efd4de8057b2af9cde7e1038c67fd0825a"
+   }
+   ```
+   
+   The response is an array of entry UUIDs for the log entries matching the digest. (Note: Rekor uses its own format for UUIDs, not IETF RFC4122 UUIDs.)
+   
+   ```json
+   ["e66b75a340d2cf240f92cc0fd7f77c3ea0855c7eb5e3e9e2ae873567d9491fdc"]
+   ```
+
+1. Retrieve the log entries by their UUIDs.
+
+   ```
+   POST https://rekor.sigstore.dev/api/v1/log/entries/retrieve
+   {
+     entryUUIDs: ["e66b75a340d2cf240f92cc0fd7f77c3ea0855c7eb5e3e9e2ae873567d9491fdc"]
+   }
+   ```
+
+   The response is an array of log entries (abbreviated), keyed by the log entry UUIDs we submitted:
+   
+   ```jsonc
+   [
+    {
+      "e66b75a340d2cf240f92cc0fd7f77c3ea0855c7eb5e3e9e2ae873567d9491fdc": {
+        "attestation": {},
+        "body": "eyJhcGlWZXJza[...]FMwdExRbz0ifX19fQ==",
+        "integratedTime": 1642790103,
+        "logID": "c0d23d6ad406973f9559f3bc5b8445c224f98b9591801d",
+        "logIndex": 1124311,
+        "verification": {
+          "inclusionProof": {
+               //…
+        	},
+          "signedEntryTimestamp": "MEUCIQDWwF3ehGj+9BBsVdnm5Bg[...]eooKBPZkXLLDVl9t72FkY/VUHw="
```

Comment:
Do you also plan to verify the Rekor entry? Note that you'll need to distribute the verification keys securely too. See comment below about TUF.

Reply:
Yes, I think the log entry should be verified, whether it's retrieved from Rekor itself or whether it's cached on rubygems.org. I don't yet know how the gem client would acquire/manage the key needed to check the signedEntryTimestamp.

As you throroughly explained in your TUF comment below, this is not trivial.

Reply:
I wouldn't suggest pulling it directly from Rekor on each verification, since you'll be trusting the log to not lie or be compromised. You could cache the key at a trusted source like rubygems.org. You'll need to figure out an update mechanism too, which minimally could be supporting multiple verification keys.

I think it's also important to verify the Fulcio certificates without fetching the root from CA issuers. The AIA information for CA issuers is meant for chain building. Also, since Rekor and Fulcio are separate systems, I wouldn't necessarily trust that the certificate put into Rekor chains up to a trusted root.